tests/kernel/smp/src/main.c

```diff
@@ -441,6 +444,105 @@ void test_wakeup_threads(void)
 	cleanup_resources();
 }
 
+/* a thread for testing SMP IPI */
+static void thread_get_cpu_entry(void *p1, void *p2, void *p3)
+{
+	int bsp_id = *(int *)p1;
+	int cpu_id = -1;
+
+	/* get current cpu number for running thread */
+	_cpu_t *curr_cpu = arch_curr_cpu();
+
+	/**TESTPOINT: call arch_curr_cpu() to get cpu struct */
+	zassert_true(curr_cpu != NULL,
+			"test failed to get current cpu.");
+
+	cpu_id = curr_cpu->id;
+
+	zassert_true(bsp_id != cpu_id,
+			"should not be the same with our BSP");
+}
+
+/**
+ * @brief Test get a pointer of CPU
+ *
+ * @ingroup kernel_smp_tests
+ *
+ * @details Architecture layer provides a mechanism to return a pointer to the
+ * current kernel CPU record of the running CPU.
+ *
+ * We call arch_curr_cpu() and get it's member, both in main and spwaned thread
+ * speratively, and compare them. They shall be different in SMP enviornment.
+ *
+ * @see arch_curr_cpu()
+ */
+void test_get_cpu(void)
+{
+	k_tid_t thread_id;
+
+	/* get current cpu number */
+	int cpu_id = arch_curr_cpu()->id;
+
+	thread_id = k_thread_create(&t2, t2_stack, T2_STACK_SIZE,
+				      (k_thread_entry_t)thread_get_cpu_entry,
+				      &cpu_id, NULL, NULL,
+				      K_PRIO_COOP(2),
+				      K_INHERIT_PERMS, K_NO_WAIT);
```

Comment:
You need to k_busy_wait() or something here to be sure that the other thread runs on a different CPU.  Once you exit the test routine the thread will terminate itself and this CPU will be free to run your test thread.  It is also possible for the test thread to exit before the child thread runs, and any zassert calls will then be ignored because the test has already passed.

In practice, a system with a working IPI will run that new thread synchronously (or almost so), but this needs to work then CONFIG_SCHED_IPI_SUPPORTED=n.

Reply:
Got it. I try to put a k_busy_wait() after spawning a thread, and make this thread run a infinite while loop. Then abort it.